### ISSUE TITLE
Add NuGet package version to ThisAssembly

### ISSUE THREAD

**Issue by qmfrederik** [AArnott/Nerdbank.GitVersioning#409]
I'm working on a .NET tool which will add a reference to a NuGet package to a project file, very similar to what `nbgv` does.

For this, I'd like to use a property on `ThisAssembly` which would return the NuGet package version.

@AArnott  what do you think?

**Comment by AArnott** [comment]
It feels a bit odd. An project might not even pack itself, so it wouldn't know the version of its containing package yet with this generated it would believe it does.

I guess I'd accept a PR with this change, but I think I'd want it opt-in via an msbuild property.

**Comment by AArnott** [comment]
Maybe it can be activated by the `IsPackable=true` property in the project file. That would make it 'right' about 99% of the time.

**Comment by grokys** [comment]
This sounds something that I could use.

For background: I have a desktop app and want to show the version in the About dialog as e.g. 3.0.0+7f41ecfa0f for a non-public build and just 3.0.0 for a public build

As far as I can see there's currently no way to detect from C# whether the build was a public release or not, is that correct?

**Comment by AArnott** [comment]
@grokys You can already tell if it's a public release since `ThisAssembly.IsPublicRelease` is a property you can access:

https://github.com/dotnet/Nerdbank.GitVersioning/blob/d7a4eb417087a0ae78bdd7ee65861375b10dab6d/src/NerdBank.GitVersioning.Tests/AssemblyInfoTest.cs#L106

The authoritative list is here:

https://github.com/dotnet/Nerdbank.GitVersioning/blob/d7a4eb417087a0ae78bdd7ee65861375b10dab6d/src/Nerdbank.GitVersioning.Tasks/AssemblyVersionInfo.cs#L162-L179

**Comment by grokys** [comment]
Hmm, strange - I'm not seeing that:

```
//------------------------------------------------------------------------------
// <auto-generated>
//     This code was generated by a tool.
//     Runtime Version:4.0.30319.42000
//
//     Changes to this file may cause incorrect behavior and will be lost if
//     the code is regenerated.
// </auto-generated>
//------------------------------------------------------------------------------

[assembly: System.Reflection.AssemblyVersionAttribute("3.0.0.0")]
[assembly: System.Reflection.AssemblyFileVersionAttribute("3.0.0.25")]
[assembly: System.Reflection.AssemblyInformationalVersionAttribute("3.0.0+1a39aac1e1")]
#if NETSTANDARD || NETFRAMEWORK || NETCOREAPP
[System.CodeDom.Compiler.GeneratedCode("Nerdbank.GitVersioning.Tasks","3.0.25.34678")]
#endif
internal static partial class ThisAssembly {
    internal const string AssemblyVersion = "3.0.0.0";
    internal const string AssemblyFileVersion = "3.0.0.25";
    internal const string AssemblyInformationalVersion = "3.0.0+1a39aac1e1";
    internal const string AssemblyName = "ABC";
    internal const string AssemblyConfiguration = "Release";
    internal const string GitCommitId = "1a39aac1e1cebff7421259e5db45aff3c1061028";
    internal static readonly System.DateTime GitCommitDate = new System.DateTime(637255600010000000L, System.DateTimeKind.Utc);
    internal const string RootNamespace = "ABC";
}
```

**Comment by grokys** [comment]
Oh wait, ignore that. Looks like NuGet wasn't showing me the newer versions of the packages. I can see it now.

**Comment by perpetualKid** [comment]
For a desktop app trying to identify if there is an (self)update available via Nuget feed, including non-public prereleases. The app only has `AssemblyInformationalVersion: 1.3.21-dev.6+631091b0`, while the `NuGetPackageVersion: 1.3.21-dev.6.g631091b0` is slightly different format in metadata vs release labels.

If there was a NuGetPackageVersion available in ThisAssembly, it would be easy to compare. 
I've been trying to use the AdditionalThisAssemblyFields `<AdditionalThisAssemblyFields Include="NuGetPackageVersion" String="???"/>` but failing to understand how to inject NuGetPackageVersion from NBGV as variable at build time. Any ideas, or did I miss some other way to compare an assembly version against a package version? @AArnott

**Comment by AArnott** [comment]
> If there was a NuGetPackageVersion available in ThisAssembly, it would be easy to compare.

Really? What compare method would you use? You'd need to know whether the version string is using semver1 or 2 rules then use a compare function appropriate for that semver syntax. It wouldn't be a simple `Version` class comparer. Would you redist a NuGet assembly and use its comparer function?

> I've been trying to use the AdditionalThisAssemblyFields

That would have *very* specific build ordering requirements. But it should be possible. Like this:

```xml
<Target Name="AddNuGetPackageVersionToThisAssembly"
        DependsOnTargets="GetBuildVersion"
        BeforeTargets="GenerateAssemblyVersionInfo">
  <ItemGroup>
    <AdditionalThisAssemblyFields Include="NuGetPackageVersion" String="$(NuGetPackageVersion)" />
  </ItemGroup>
</Target>
```

**Comment by perpetualKid** [comment]
> > If there was a NuGetPackageVersion available in ThisAssembly, it would be easy to compare.
> 
> Really? What compare method would you use? You'd need to know whether the version string is using semver1 or 2 rules then use a compare function appropriate for that semver syntax. It wouldn't be a simple `Version` class comparer. Would you redist a NuGet assembly and use its comparer function?

correct, using Nuget.Versioning to provide SemVer-compatible version comparison, to enable auto-update for the application itself, and (future plan) application content, aside using Nuget Core/Protocolto handle feed, download and unpacking. 
All worked well from the ground for public releases, just got stuck when offering an additional update channel for "developer versions" which are basically non-public releases in nbgv semantic (and served of a separate feed). Certainly not a standard use case.

> 
> > I've been trying to use the AdditionalThisAssemblyFields
> 
> That would have _very_ specific build ordering requirements. But it should be possible. Like this:
> 
> ```
> <Target Name="AddNuGetPackageVersionToThisAssembly"
>         DependsOnTargets="GetBuildVersion"
>         BeforeTargets="GenerateAssemblyVersionInfo">
>   <ItemGroup>
>     <AdditionalThisAssemblyFields Include="NuGetPackageVersion" String="$(NuGetPackageVersion)" />
>   </ItemGroup>
> </Target>
> ```

aweseome just did a quick test and seems to do that I needed, saves me from string-mangeling metadata into release labels for non-public releases. Thanks for quick response, and great tool!

**Comment by perpetualKid** [comment]
> > > If there was a NuGetPackageVersion available in ThisAssembly, it would be easy to compare.
> > 
> > 
> > Really? What compare method would you use? You'd need to know whether the version string is using semver1 or 2 rules then use a compare function appropriate for that semver syntax. It wouldn't be a simple `Version` class comparer. Would you redist a NuGet assembly and use its comparer function?
> 
> correct, using Nuget.Versioning to provide SemVer-compatible version comparison, to enable auto-update for the application itself, and (future plan) application content, aside using Nuget Core/Protocolto handle feed, download and unpacking. All worked well from the ground for public releases, just got stuck when offering an additional update channel for "developer versions" which are basically non-public releases in nbgv semantic (and served of a separate feed). Certainly not a standard use case.
> 
> > > I've been trying to use the AdditionalThisAssemblyFields
> > 
> > 
> > That would have _very_ specific build ordering requirements. But it should be possible. Like this:
> > ```
> > <Target Name="AddNuGetPackageVersionToThisAssembly"
> >         DependsOnTargets="GetBuildVersion"
> >         BeforeTargets="GenerateAssemblyVersionInfo">
> >   <ItemGroup>
> >     <AdditionalThisAssemblyFields Include="NuGetPackageVersion" String="$(NuGetPackageVersion)" />
> >   </ItemGroup>
> > </Target>
> > ```
> 
> aweseome just did a quick test and seems to do that I needed, saves me from string-mangeling metadata into release labels for non-public releases. Thanks for quick response, and great tool!



> > If there was a NuGetPackageVersion available in ThisAssembly, it would be easy to compare.
> 
> Really? What compare method would you use? You'd need to know whether the version string is using semver1 or 2 rules then use a compare function appropriate for that semver syntax. It wouldn't be a simple `Version` class comparer. Would you redist a NuGet assembly and use its comparer function?
> 
> > I've been trying to use the AdditionalThisAssemblyFields
> 
> That would have _very_ specific build ordering requirements. But it should be possible. Like this:
> 
> ```
> <Target Name="AddNuGetPackageVersionToThisAssembly"
>         DependsOnTargets="GetBuildVersion"
>         BeforeTargets="GenerateAssemblyVersionInfo">
>   <ItemGroup>
>     <AdditionalThisAssemblyFields Include="NuGetPackageVersion" String="$(NuGetPackageVersion)" />
>   </ItemGroup>
> </Target>
> ```

somehow after 3.4.255 this does not seem to work anymore, in 3.5.103+ it the additional field is not created anymore, breaking our code (error CS0117: 'ThisAssembly' does not contain a definition for 'NuGetPackageVersion'). Any idea @AArnott ?

**Comment by AArnott** [comment]
Thanks for sharing. The target names had to be changed IIRC to avoid conflicting with a target from another SDK.

**Comment by AArnott** [comment]
@wjrogers Can you report your assembly informational version to the error tracking service? That always includes the git commit id. From that you could derive the nuget package version while debugging, if the two aren't already equal.

**Comment by wjrogers** [comment]
Yes, I suppose, but I'd rather have everything match the release version for clarity.
